File: scripts/seed-supabase.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { createClient } from '../frontend/node_modules/@supabase/supabase-js/dist/index.mjs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load root .env file manually
const envPath = path.join(__dirname, '../.env');
if (fs.existsSync(envPath)) {
  const envContent = fs.readFileSync(envPath, 'utf-8');
  envContent.split('\n').forEach(line => {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith('#')) {
      const [key, ...vals] = trimmed.split('=');
      if (key && vals.length > 0) {
        const val = vals.join('=').trim().replace(/^["']|["']$/g, '');
        process.env[key.trim()] = val;
      }
    }
  });
}

// Read Environment Variables
const supabaseUrl = process.env.SUPABASE_URL || process.env.VITE_SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.SUPABASE_ANON_KEY || process.env.VITE_SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseKey || supabaseUrl.includes('placeholder')) {
  console.error('❌ Error: VITE_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY (or VITE_SUPABASE_ANON_KEY) must be properly set in .env');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

// Robust CSV Parser handling quoted fields and line breaks
function parseCSV(content) {
  const lines = [];
  let currentField = '';
  let inQuotes = false;
  let currentLine = [];

  for (let i = 0; i < content.length; i++) {
    const char = content[i];
    const nextChar = content[i + 1];

    if (char === '"') {
      if (inQuotes && nextChar === '"') {
        currentField += '"';
        i++; // skip escaped quote
      } else {
        inQuotes = !inQuotes;
      }
    } else if (char === ',' && !inQuotes) {
      currentLine.push(currentField);
      currentField = '';
    } else if ((char === '\r' || char === '\n') && !inQuotes) {
      if (char === '\r' && nextChar === '\n') i++;
      currentLine.push(currentField);
      if (currentLine.some(f => f.length > 0)) {
        lines.push(currentLine);
      }
      currentLine = [];
      currentField = '';
    } else {
      currentField += char;
    }
  }

  if (currentField.length > 0 || currentLine.length > 0) {
    currentLine.push(currentField);
    lines.push(currentLine);
  }

  if (lines.length === 0) return [];
  const headers = lines[0].map(h => h.trim());
  
  return lines.slice(1).map(line => {
    const obj = {};
    headers.forEach((h, idx) => {
      let val = line[idx] !== undefined ? line[idx].trim() : null;
      if (val === '' || val === 'NULL' || val === 'null') {
        val = null;
      }
      obj[h] = val;
    });
    return obj;
  });
}

// Transform row values according to table type rules
function transformRow(tableName, row) {
  const transformed = { ...row };

  // Helper for booleans
  const toBool = (val) => {
    if (val === null || val === undefined) return false;
    return val === '1' || val === 'true' || val === 'TRUE' || val === true;
  };

  // Helper for numbers
  const toNum = (val) => (val !== null && val !== undefined && val !== '') ? Number(val) : null;

  switch (tableName) {
    case 'languages':
      transformed.rtl = toBool(row.rtl);
      transformed.tts_supported = toBool(row.tts_supported);
      break;
    case 'currencies':
      transformed.minor_unit_exponent = toNum(row.minor_unit_exponent);
      break;
    case 'fx_rates':
      transformed.rate = toNum(row.rate);
      break;
    case 'receipts':
      transformed.total_amount_truth = toNum(row.total_amount_truth);
      if (typeof row.line_items_truth === 'string') {
        try {
          transformed.line_items_truth = JSON.parse(row.line_items_truth);
        } catch (e) {
          transformed.line_items_truth = [];
        }
      }
      break;
    case 'cities':
      transformed.lat = toNum(row.lat);
      transformed.lng = toNum(row.lng);
      transformed.population = toNum(row.population);
      break;
    case 'trips':
      transformed.party_size = toNum(row.party_size);
      transformed.adults = toNum(row.adults);
      transformed.children = toNum(row.children);
      transformed.is_group_trip = toBool(row.is_group_trip);
      break;
    case 'budgets':
      transformed.total_amount = toNum(row.total_amount);
      transformed.accommodation_cap = toNum(row.accommodation_cap);
      transformed.transport_cap = toNum(row.transport_cap);
      transformed.food_cap = toNum(row.food_cap);
      transformed.activities_cap = toNum(row.activities_cap);
      transformed.misc_cap = toNum(row.misc_cap);
      transformed.alert_threshold_pct = toNum(row.alert_threshold_pct);
      break;
    case 'expenses':
      transformed.amount = toNum(row.amount);
      transformed.home_amount = toNum(row.home_amount);
      transformed.is_settled = toBool(row.is_settled);
      break;
    case 'itineraries':
      transformed.version = toNum(row.version);
      transformed.is_active = toBool(row.is_active);
      transformed.total_cost = toNum(row.total_cost);
      transformed.total_duration_minutes = toNum(row.total_duration_minutes);
      transformed.total_carbon_kg = toNum(row.total_carbon_kg);
      if (typeof row.optimizer_weights === 'string' && row.optimizer_weights) {
        try {
          transformed.optimizer_weights = JSON.parse(row.optimizer_weights);
        } catch (e) {
          transformed.optimizer_weights = null;
        }
      }
      break;
    case 'itinerary_items':
      transformed.day_index = toNum(row.day_index);
      transformed.sort_order = toNum(row.sort_order);
      transformed.cost = toNum(row.cost);
      transformed.carbon_kg = toNum(row.carbon_kg);
      transformed.duration_minutes = toNum(row.duration_minutes);
      transformed.locked = toBool(row.locked);
      break;
    case 'trip_members':
      transformed.share_weight = toNum(row.share_weight);
      break;
    case 'expense_splits':
      transformed.share_value = toNum(row.share_value);
      transformed.amount = toNum(row.amount);
      break;
  }

  return transformed;
}

// 15 Canonical Tables in strict dependency order
const tables = [
  { file: '01_categories.csv', table: 'categories' },
  { file: '02_currencies.csv', table: 'currencies' },
  { file: '03_fx_rates.csv', table: 'fx_rates' },
  { file: '04_languages.csv', table: 'languages' },
  { file: '05_receipts.csv', table: 'receipts' },
  { file: '06_countries.csv', table: 'countries' },
  { file: '07_cities.csv', table: 'cities' },
  { file: '08_users.csv', table: 'users' },
  { file: '09_trips.csv', table: 'trips' },
  { file: '10_budgets.csv', table: 'budgets' },
  { file: '11_expenses.csv', table: 'expenses' },
  { file: '12_itineraries.csv', table: 'itineraries' },
  { file: '13_itinerary_items.csv', table: 'itinerary_items' },
  { file: '14_trip_members.csv', table: 'trip_members' },
  { file: '15_expense_splits.csv', table: 'expense_splits' }
];

async function runMigration() {
  console.log('🚀 Starting Supabase Migration for 15 Canonical PS-08 Tables...\n');
  const csvDir = path.join(__dirname, '../data-model/seed/csv');

  const summary = [];

  for (const { file, table } of tables) {
    const filePath = path.join(csvDir, file);
    if (!fs.existsSync(filePath)) {
      console.warn(`⚠️ Warning: CSV file not found: ${file}`);
      continue;
    }

    const content = fs.readFileSync(filePath, 'utf-8');
    const rawRows = parseCSV(content);
    const transformedRows = rawRows.map(row => transformRow(table, row));

    console.log(`📦 Upserting ${transformedRows.length} rows into table '${table}'...`);

    const BATCH_SIZE = 500;
    let insertedCount = 0;

    for (let i = 0; i < transformedRows.length; i += BATCH_SIZE) {
      const chunk = transformedRows.slice(i, i + BATCH_SIZE);
      const { data, error } = await supabase.from(table).upsert(chunk, { ignoreDuplicates: false });

      if (error) {
        console.error(`❌ Error upserting into '${table}' (batch ${i}):`, error.message);
        throw error;
      }
      insertedCount += chunk.length;
    }

    console.log(`✅ Successfully imported ${insertedCount} rows into '${table}'.\n`);
    summary.push({ table, count: insertedCount });
  }

  console.log('----------------------------------------------------');
  console.log('🎉 MIGRATION COMPLETE! SUMMARY:');
  console.table(summary);
  const totalRows = summary.reduce((acc, curr) => acc + curr.count, 0);
  console.log(`Total Rows Imported: ${totalRows}`);
}

runMigration().catch(err => {
  console.error('Fatal migration error:', err);
  process.exit(1);
});

File: scripts/seed-supabase.py
import os
import json
import sqlite3
import urllib.request
import urllib.error

# 1. Read .env file manually
env_path = os.path.join(os.path.dirname(__file__), '../.env')
if os.path.exists(env_path):
    with open(env_path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith('#'):
                parts = line.split('=', 1)
                if len(parts) == 2:
                    k, v = parts[0].strip(), parts[1].strip().strip("\"'")
                    os.environ[k] = v

supabase_url = os.getenv('SUPABASE_URL') or os.getenv('VITE_SUPABASE_URL')
supabase_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY') or os.getenv('SUPABASE_ANON_KEY') or os.getenv('VITE_SUPABASE_ANON_KEY')

if not supabase_url or not supabase_key or 'placeholder' in supabase_url:
    print("❌ Error: Valid VITE_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY (or VITE_SUPABASE_ANON_KEY) must be set in .env")
    exit(1)

db_path = os.path.join(os.path.dirname(__file__), '../data-model/seed/PS-08.db')
if not os.path.exists(db_path):
    print(f"❌ Error: Database file not found at {db_path}")
    exit(1)

conn = sqlite3.connect(db_path)
conn.row_factory = sqlite3.Row
cur = conn.cursor()

tables = [
    'categories', 'currencies', 'fx_rates', 'languages', 'receipts',
    'countries', 'cities', 'users', 'trips', 'budgets',
    'expenses', 'itineraries', 'itinerary_items', 'trip_members', 'expense_splits'
]

def to_int_bool(val):
    if val is None:
        return 0
    return 1 if val in (1, '1', 'true', 'TRUE', True) else 0

def transform_row(table_name, row_dict):
    d = dict(row_dict)
    
    # Process booleans and JSONs according to database schema
    if table_name == 'languages':
        d['rtl'] = to_int_bool(d.get('rtl'))
        d['tts_supported'] = to_int_bool(d.get('tts_supported'))
    elif table_name == 'receipts':
        if isinstance(d.get('line_items_truth'), str):
            try:
                d['line_items_truth'] = json.loads(d['line_items_truth'])
            except:
                d['line_items_truth'] = []
    elif table_name == 'trips':
        d['is_group_trip'] = to_int_bool(d.get('is_group_trip'))
    elif table_name == 'expenses':
        d['is_settled'] = to_int_bool(d.get('is_settled'))
    elif table_name == 'itineraries':
        d['is_active'] = to_int_bool(d.get('is_active'))
        if isinstance(d.get('optimizer_weights'), str) and d['optimizer_weights']:
            try:
                d['optimizer_weights'] = json.loads(d['optimizer_weights'])
            except:
                d['optimizer_weights'] = None
    elif table_name == 'itinerary_items':
        d['locked'] = to_int_bool(d.get('locked'))
        
    return d

def batch_upsert(table_name, rows):
    url = f"{supabase_url.rstrip('/')}/rest/v1/{table_name}"
    headers = {
        'apikey': supabase_key,
        'Authorization': f"Bearer {supabase_key}",
        'Content-Type': 'application/json',
        'Prefer': 'resolution=merge-duplicates'
    }
    
    data_bytes = json.dumps(rows).encode('utf-8')
    req = urllib.request.Request(url, data=data_bytes, headers=headers, method='POST')
    
    try:
        with urllib.request.urlopen(req) as resp:
            return resp.status in (200, 201, 204)
    except urllib.error.HTTPError as e:
        err_msg = e.read().decode('utf-8')
        print(f"❌ HTTP Error for table '{table_name}': {e.code} {e.reason} - {err_msg}")
        return False
    except Exception as e:
        print(f"❌ Request Error for table '{table_name}': {e}")
        return False

print("🚀 Starting Supabase Database Seeding from PS-08.db...\n")

total_imported = 0
summary = []

for table in tables:
    cur.execute(f"SELECT * FROM {table}")
    rows = [transform_row(table, row) for row in cur.fetchall()]
    
    print(f"📦 Upserting {len(rows)} rows into table '{table}'...")
    
    batch_size = 500
    success_count = 0
    for i in range(0, len(rows), batch_size):
        chunk = rows[i:i + batch_size]
        if batch_upsert(table, chunk):
            success_count += len(chunk)
        else:
            print(f"Failed to upsert batch for table {table}")
            exit(1)
            
    print(f"✅ Successfully imported {success_count} rows into '{table}'.\n")
    summary.append((table, success_count))
    total_imported += success_count

print("----------------------------------------------------")
print("🎉 MIGRATION COMPLETE! SUMMARY:")
for t, c in summary:
    print(f"  • {t}: {c} rows")
print(f"\nTotal Rows Imported into Supabase: {total_imported}")
